docs/design/distributed-cache.md

```diff
@@ -0,0 +1,110 @@
+
+## Goals
+
+1. **Scalability**: Distributed caches can be used to distribute the load across multiple instances of an application. As the application scales, the cache can help maintain consistent performance by providing a shared data store that all instances can access.
+2. **Improved Performance**: By sharing cache accross multiple JSON-RPC Relay instances, correct data will always be available for methods to access upon requests.
+3. **Reduced Load on Mirror-Node**: By caching data and serving it directly from the cache, the load on the underlying data sources (such as Mirror-Node) is reduced. This can help prevent performance bottlenecks and overloading of these resources.
+
+## Architecture
+
+For distributed cache JSON-RPC Relay will be using Redis and in order to establish connection and manage, it needs to utilize `node-redis` package. All cache managment will be done in the respective cache client, either for the internal `lru-cache` or shared redis, which methods will be defined by an interface. Using dependency injection pattern with dependency inversion all services that require cache will receive only abstraction in the constructor. This will allow for flexible usage of the cache without additional coupling, depending on the service need.
+Important details is that, if an operator does not want to use Redis, it does not need to, because all services that use shared cache will automatically switch to using internal `lru-cache`.
+
+### Diagram
+
+```mermaid
+flowchart TD
+    subgraph Relay Instance 1
+        subgraph Services Instance 1
+        A{{Filter Service}}
+        B{{Websocket Service}}
+        C{{History Service}}
+        end
+            subgraph Internal Cache Instance 1
+            D{{lru-cache}}
+            end
+    end
+    subgraph Distributed Cache
+    I{{Redis}}
+    end
+    subgraph Relay Instance 2
+        subgraph Services Instance 2
+        E{{Filter Service}}
+        F{{Websocket Service}}
+        G{{History Service}}
+        end
+            subgraph Internal Cache Instance 2
+            H{{lru-cache}}
+            end
+    end
+    A == Redis is enabled ==> I
+    E == Redis is enabled ==> I
+    A == Redis is disabled ==> D
+    E == Redis is disabled ==> H
+    B ==> D
+    C ==> D
+    F ==> H
+    G ==> H
+```
+
+### Interface
+
+```javascript
+interface ICacheClient {
+  get(key: string, callingMethod: string, requestIdPrefix?: string): any;
+  set(key: string, value: any, callingMethod: string, ttl?: number, requestIdPrefix?: string): void;
+  delete(key: string, callingMethod: string, requestIdPrefix?: string): void;
+  purgeStale(): void;
+  clear(): void;
+}
+```
+
+## Limits
+
+1. Should distributed cache be enabled. Env. variable `REDIS_ENABLED`.
+
+## Metric Capturing
+
+Capture metrics for the following:
+
+1. Log every service what kind of cache it's using.
+2. For each service log calls to `get`, `set`, `delete`.
+
```

Comment:
This is related to the comment at the end, I thought that the `cacheType` only tells us if is lru(local) or redis (distributed). 

But I think it would be useful to know when the cache is being used, for state storage or caching purposes.
The rationale being that caching would not disrupt the features if gets deleted/cleared, while the state storage function would (filterId or synthetic transactions). 

So not only would be a nice tracking label, but even if the key could have a prefix would make easier to delete/clear any or all caches, without affecting features that need to share state between pods.